subprojects/execution/src/main/java/org/gradle/internal/execution/history/impl/OutputSnapshotUtil.java

```diff
@@ -48,14 +48,17 @@
 public class OutputSnapshotUtil {
 
     /**
-     * Filters out snapshots that are not considered outputs. Entries that are considered outputs are:
-     * <ul>
-     * <li>an entry that did not exist before the execution, but exists after the execution</li>
-     * <li>an entry that did exist before the execution, and has been changed during the execution</li>
-     * <li>an entry that did wasn't changed during the execution, but was already considered an output during the previous execution</li>
-     * </ul>
+     * Finds outputs that are still present since the last execution when overlapping outputs are present.
+     *
+     * Note: when there are no overlapping outputs, all outputs currently existing in the output locations
+     * are considered outputs of the work.
+     *
+     * If a property did not exist after the previous execution then all the outputs for it will be ignored.
```

Comment:
Why do we filter in this case? Is that because the filtering doesn't happen lazily any more? We shouldn't query output changes when the set of output properties has changed. Can we easily remove this special case by moving the filtering into the accept method of `OutputFileChanges`?

Reply:
We currently consider adding a `null` output property a no change. We could change this behavior, but that'd be a slightly breaking change perhaps. I don't think we should make that change in this PR.